ReactiveCocoa/DelegateProxy.swift

```diff
@@ -1,106 +1,262 @@
 import ReactiveSwift
 import enum Result.NoError
 
-internal class DelegateProxy<Delegate: NSObjectProtocol>: NSObject {
-	internal weak var forwardee: Delegate? {
-		didSet {
-			originalSetter(self)
-		}
+private let delegateProxySetupKey = AssociationKey<Bool>(default: false)
+private let hasSwizzledKey = AssociationKey<Bool>(default: false)
+
+internal protocol DelegateProxyProtocol: class {
+	var _forwardee: AnyObject? { get }
+	func runtimeWillIntercept(_ selector: Selector, signal: Signal<AnyObject, NoError>) -> Signal<AnyObject, NoError>
+}
+
+/// An opaque delegate proxy configuration.
+public struct DelegateProxyConfiguration {
+	fileprivate let objcProtocol: Protocol
+	fileprivate let lifetime: Lifetime
+	fileprivate let originalSetter: (AnyObject) -> Void
+}
+
+public class DelegateProxy<Delegate: NSObjectProtocol>: NSObject, DelegateProxyProtocol {
```

Comment:
Why can't `DelegateProxy` be `final`?

Reply:
https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3467#discussion_r136692297

But if you still think it is more sensible for us to start with `final`, I am fine with that.

Reply:
But if we made it `final`, then we could make [this method private](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3467/files#r250601283)? That seems worth doing

Reply:
Yep.

Reply:
In any case I think intercepting non-void returning methods is probably very niche, and a rather dangerous zone to walk into with little return. So closing it should be fine.